File: Sources/Application/SharedCoreTrafficShadow.swift
import os

enum SharedCoreTrafficShadow {
  typealias Reporter = @Sendable (SharedCoreTrafficShadowObservation) -> Void

  private struct State {
    var reporter: Reporter?
    var observationGate = SharedCoreTrafficShadowObservationGate()
  }

  private static let stateLock = OSAllocatedUnfairLock(initialState: State())

  static func configure(reporter: Reporter?) {
    stateLock.withLock { state in
      state.reporter = reporter
      state.observationGate.reset()
    }
  }

  static func observe(
    bytes: UInt64,
    nativeText: String,
    format: SharedCoreTrafficFormat
  ) -> String {
    let status = SharedCoreTrafficShadowComparator.compare(
      bytes: bytes,
      nativeText: nativeText,
      format: format
    )
    let observation = SharedCoreTrafficShadowObservation(
      format: format,
      status: status
    )
    let reporter = stateLock.withLock { state -> Reporter? in
      guard let reporter = state.reporter,
        state.observationGate.shouldReport(observation)
      else {
        return nil
      }
      return reporter
    }
    reporter?(observation)
    return nativeText
  }
}

File: platform/windows/MihomoMeter.Windows.Core/Application/SharedCoreTrafficShadow.cs
namespace MihomoMeter.Windows.Core.Application;

public enum SharedCoreTrafficFormat
{
    ByteCount,
    Rate,
    CompactRate,
}

public enum SharedCoreTrafficShadowStatus
{
    Matched,
    AbiMismatch,
    NativeCallFailed,
    UnexpectedResult,
    Mismatch,
    UnknownFailure,
}

public readonly record struct SharedCoreTrafficShadowObservation(
    SharedCoreTrafficFormat Format,
    SharedCoreTrafficShadowStatus Status);

internal sealed class SharedCoreTrafficShadowObservationGate
{
    private readonly HashSet<SharedCoreTrafficShadowObservation> _reportedObservations = [];

    public bool ShouldReport(SharedCoreTrafficShadowObservation observation)
    {
        return _reportedObservations.Add(observation);
    }

    public void Reset()
    {
        _reportedObservations.Clear();
    }
}

public static class SharedCoreTrafficShadow
{
    private static readonly object StateLock = new();
    private static readonly SharedCoreTrafficShadowObservationGate ObservationGate = new();
    private static Action<SharedCoreTrafficShadowObservation>? _reporter;

    public static void ConfigureReporter(
        Action<SharedCoreTrafficShadowObservation>? reporter)
    {
        lock (StateLock)
        {
            _reporter = reporter;
            ObservationGate.Reset();
        }
    }

    public static string Observe(
        ulong bytes,
        string nativeText,
        SharedCoreTrafficFormat format)
    {
        var status = SharedCoreTrafficShadowComparator.Compare(
            bytes,
            nativeText,
            format);
        var observation = new SharedCoreTrafficShadowObservation(format, status);
        Action<SharedCoreTrafficShadowObservation>? reporter;
        lock (StateLock)
        {
            reporter = _reporter is not null && ObservationGate.ShouldReport(observation)
                ? _reporter
                : null;
        }
        try
        {
            reporter?.Invoke(observation);
        }
        catch
        {
            // 影子诊断不得影响仍由原生算法决定的生产输出。
        }
        return nativeText;
    }
}
